crates/package_manager/src/add.rs

```diff
@@ -0,0 +1,96 @@
+use crate::Install;
+use derive_more::{Display, Error};
+use miette::Diagnostic;
+use pacquet_lockfile::Lockfile;
+use pacquet_npmrc::Npmrc;
+use pacquet_package_json::PackageJsonError;
+use pacquet_package_json::{DependencyGroup, PackageJson};
+use pacquet_registry::{PackageTag, PackageVersion};
+use pacquet_tarball::Cache;
+use reqwest::Client;
+
+/// This subroutine does everything `pacquet add` is supposed to do.
+#[must_use]
+pub struct Add<'a, ListDependencyGroups, DependencyGroupList>
+where
+    ListDependencyGroups: Fn() -> DependencyGroupList,
+    DependencyGroupList: IntoIterator<Item = DependencyGroup>,
+{
+    /// Shared cache that store downloaded tarballs.
+    pub tarball_cache: &'a Cache,
+    /// HTTP client to make HTTP requests.
+    pub http_client: &'a Client,
+    /// Configuration read from `.npmrc`.
+    pub config: &'static Npmrc,
+    /// Data from the `package.json` file.
+    pub package_json: &'a mut PackageJson,
+    /// Data from the `pnpm-lock.yaml` file.
+    pub lockfile: Option<&'a Lockfile>,
+    /// List of [`DependencyGroup`]s.
+    pub list_dependency_groups: ListDependencyGroups,
+    /// Name of the package to add.
+    pub package: &'a str,
+    /// Whether `--save-exact` is provided.
+    pub save_exact: bool,
+}
+
+/// Error type of [`Add`].
+#[derive(Debug, Display, Error, Diagnostic)]
+pub enum AddError {
+    #[display("Failed to add package to manifest: {_0}")]
+    AddDependencyToPackageJson(#[error(source)] PackageJsonError),
+    #[display("Failed save the manifest file: {_0}")]
+    SavePackageJson(#[error(source)] PackageJsonError),
+}
+
+impl<'a, ListDependencyGroups, DependencyGroupList>
+    Add<'a, ListDependencyGroups, DependencyGroupList>
+where
+    ListDependencyGroups: Fn() -> DependencyGroupList,
+    DependencyGroupList: IntoIterator<Item = DependencyGroup>,
+{
+    pub async fn run(self) -> Result<(), AddError> {
```

Comment:
Can't we just use the struct as options in a `addDependencies` function?

Reply:
What exactly do you mean? What's the struct you were referring to?

Reply:
I mean I don't know, maybe this is the widespread pattern used in Rust but in other languages I guess I wouldn't create a class just for a single method.

You could do

```
struct AddOptions {
  ...
}

pub async fn addDependencies (opts: AddOptions) -> Result<(), AddError> {
  ...
}
```

Reply:
It can work:

```diff
diff --git a/crates/cli/src/cli_args/add.rs b/crates/cli/src/cli_args/add.rs
index 1189ae2..c324032 100644
--- a/crates/cli/src/cli_args/add.rs
+++ b/crates/cli/src/cli_args/add.rs
@@ -1,7 +1,7 @@
 use crate::State;
 use clap::Args;
 use miette::Context;
-use pacquet_package_manager::Add;
+use pacquet_package_manager::{add_dependency, AddDependencyOptions as AddOptions};
 use pacquet_package_manifest::DependencyGroup;
 use std::path::PathBuf;
 
@@ -87,7 +87,7 @@ impl AddArgs {
 
         let State { tarball_cache, http_client, config, manifest, lockfile } = &mut state;
 
-        Add {
+        add_dependency(AddOptions {
             tarball_cache,
             http_client,
             config,
@@ -96,8 +96,7 @@ impl AddArgs {
             list_dependency_groups: || self.dependency_options.dependency_groups(),
             package: &self.package,
             save_exact: self.save_exact,
-        }
-        .run()
+        })
         .await
         .wrap_err("adding a new package")
     }
```

But now the consumer have to import 2 items (the function and the options type) instead of just 1 previously (just the subroutine). Not to mention, in order to make the code consistent, I would have to change the other subroutines to follow similar pattern too (which would consume a lot of time).

Reply:
Is it considered to be best practice in Rust?

It is for sure better than it was previously with many arguments. Just not usual for me. I haven't seen a lot of rust code. What I've seen mostly used functions with many args or a more OOP styled approach.

Reply:
It is unfortunate that [named arguments](https://internals.rust-lang.org/t/pre-rfc-named-arguments/16413) doesn't yet exist in Rust.

> Is it considered to be best practice in Rust?

The convention is actually builder pattern. So it would look something like this:

```rust
Add::builder()
    .tarball_cache(tarball_cache)
    .http_client(http_client)
    // ...
    .save_exact()
    .build()
    .run()
    .await
    .wrap_err("adding a new package")
```

The builder pattern is already used in `std` too. I believe you must have seen this before:

```rust
let mut file = OpenOptions::new().create(true).write(true).open()?;
```

The reason I didn't apply it is because I don't want the code to take too much time to compile.

Besides, my way, your way, and even the builder pattern way are all but different syntax to do the same thing anyway.

Reply:
I won't dictate my way at the moment as I don't feel I have enough knowledge of rust yet. In any case, your way looks better to me than the builder pattern.